File: Algorithms/bfs.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

vector<vector<int>> adjacencyList;

vector<vector<int>> bfs(vector<vector<int>>& adjacencyList, int numNodes, int source) {
    vector<bool> visited(numNodes, false);
    vector<int> distances(numNodes), parents(numNodes);
    queue<int> queue;

    queue.push(source);
    visited[source] = true;
    distances[source] = 0;
    parents[source] = -1;

    while (!queue.empty()) {
        int node = queue.front();
        queue.pop();
        cout << node << endl;

        for (auto neighbor : adjacencyList[node]) {
            if (!visited[neighbor]) {
                queue.push(neighbor);
                visited[neighbor] = true;
                distances[neighbor] = distances[node] + 1;
                parents[neighbor] = node;
            }
        }
    }

    return {distances, parents};
}

void printPath(const vector<int>& parents, int target) {
    vector<int> path;

    if (parents[target] == -1) {
        cout << "No path!";
    } else {
        while (target != -1) {
            path.push_back(target);
            target = parents[target];
        }
        reverse(path.begin(), path.end());
        cout << "Path: ";
        for (int node : path) {
            cout << node << " ";
        }
    }
    cout << endl;
}

int main() {
    int numNodes;
    int source;

    vector<int> distances;
    vector<int> parents;

    tie(distances, parents) = bfs(adjacencyList, numNodes, source);

    int target;
    printPath(parents, target);

    return 0;
}

File: Algorithms/bfs.py
from collections import deque

def bfs(adj_list, num_nodes, source):
    visited = [False] * num_nodes
    distances = [0] * num_nodes
    parents = [-1] * num_nodes
    queue = deque()

    queue.append(source)
    visited[source] = True
    distances[source] = 0

    while queue:
        node = queue.popleft()
        print(node)

        for neighbor in adj_list[node]:
            if not visited[neighbor]:
                queue.append(neighbor)
                visited[neighbor] = True
                distances[neighbor] = distances[node] + 1
                parents[neighbor] = node

    return distances, parents

def print_path(parents, target):
    if not visited[target]:
        print("No path!")
    else:
        path = []
        while target != -1:
            path.append(target)
            target = parents[target]
        path.reverse()
        print("Path:", end=" ")
        print(*path)

num_nodes = 0  # number of nodes
source = 0  # source

adj_list = [[]]  # adjacency list representation

distances, parents = bfs(adj_list, num_nodes, source)

target_node = 0  # node for which you want to print the path
print_path(parents, target_node)
